webapi/src/main/scala/org/knora/webapi/routing/Authenticator.scala

```diff
@@ -462,39 +600,117 @@ object Authenticator {
-            }
+            val userProfileV1Future = for {
+                maybeUserProfileV1 <- (responderManager ? UserProfileByEmailGetV1(email, UserProfileTypeV1.FULL)).mapTo[Option[UserProfileV1]]
+                userProfileV1 = maybeUserProfileV1 match {
+                    case Some(up) => up
+                    case None => {
+                        log.debug(s"getUserProfileV1ByEmail - supplied email not found - throwing exception")
+                        throw BadCredentialsException(s"$BAD_CRED_USER_NOT_FOUND")
+                    }
+                }
+                _ = log.debug(s"getUserProfileV1ByEmail - userProfileV1: $userProfileV1")
+            } yield userProfileV1
+
+            // TODO: return the future here instead of using Await.
+            Await.result(userProfileV1Future, Duration(3, SECONDS))
+        } else {
+            throw BadCredentialsException(BAD_CRED_EMAIL_NOT_SUPPLIED)
+        }
+    }
+}
+
+object JWTHelper {
+
+    import Authenticator.AUTHENTICATION_INVALIDATION_CACHE_NAME
+
+    val algorithm = Algorithm.HS256
+    val requiredHeaders: Set[HeaderField] = Set[HeaderField](Typ)
+    val requiredClaims: Set[ClaimField] = Set[ClaimField](Iss, Sub, Aud, Iat, Exp, Jti)
+
+    val log = Logger(LoggerFactory.getLogger(this.getClass))
+
+    /**
+      * Create a JWT.
+      *
+      * @param userIri   the user IRI that will be encoded into the token.
+      * @param secretKey the secret key used for encoding.
+      * @param longevity the token's longevity in days.
+      * @return a [[String]] containg the JWT.
+      */
+    def createToken(userIri: IRI, secretKey: String, longevity: Long): String = {
+
+        val headers = Seq[HeaderValue](Typ("JWT"), Alg(algorithm))
+
+        val now: Long = System.currentTimeMillis() / 1000l
+        val nowPlusLongevity: Long = now + longevity * 60 * 60 * 24
+
+        val identifier: String = UUID.randomUUID().toString()
+
+        val claims = Seq[ClaimValue](Iss("webapi"), Sub(userIri), Aud("webapi"), Iat(now), Exp(nowPlusLongevity), Jti(identifier))
+
+        val jwt = new DecodedJwt(headers, claims)
+
+        jwt.encodedAndSigned(secretKey)
+    }
+
+    /**
+      * Validates a JWT by also taking the invalidation cache into account. The invalidation cache holds invalidated
+      * tokens, which would otherwise validate.
+      *
+      * @param token  the JWT.
+      * @param secret the secret used to encode the token.
+      * @return a [[Boolean]].
+      */
+    def validateToken(token: String, secret: String): Boolean = {
+
+        if (CacheUtil.get[UserProfileV1](AUTHENTICATION_INVALIDATION_CACHE_NAME, token).nonEmpty) {
+            // token invalidated so no need to decode
+            log.debug("validateToken - token found in invalidation cache, so not valid")
+            false
         } else {
-            throw BadCredentialsException(BAD_CRED_USERNAME_NOT_SUPPLIED)
+            DecodedJwt.validateEncodedJwt(
```

Comment:
Please used named parameters in method calls with more than one parameter, it's less error-prone and is easier to read.

Reply:
done